src/time_utils.c

```diff
@@ -125,3 +126,271 @@ ts_time_value_from_arg(Datum arg, Oid argtype, Oid timetype)
 
 	return ts_time_value_to_internal(arg, argtype);
 }
+
+/*
+ * Try to coerce a type to a supported time type.
+ *
+ * To support custom time types in hypertables, we need to know the type's
+ * boundaries in order to, e.g., construct dimensional chunk constraints. If
+ * the type is IMPLICITLY coercible to one of the supported time types we can
+ * just use the boundaries of the supported type. Thus, the custom time type
+ * will inherit the time boundaries of the first compatible time type found.
+ */
+static Oid
+coerce_to_time_type(Oid type)
+{
+	static const Oid supported_time_types[] = { DATEOID, TIMESTAMPOID, TIMESTAMPTZOID,
+												INT2OID, INT4OID,	  INT8OID };
+
+	int i;
+
+	for (i = 0; i < lengthof(supported_time_types); i++)
+	{
+		Oid targettype = supported_time_types[i];
+
+		if (can_coerce_type(1, &type, &targettype, COERCION_IMPLICIT))
+			return targettype;
+	}
+
+	elog(ERROR, "unsupported time type \"%s\"", format_type_be(type));
+	pg_unreachable();
+}
+
+/*
+ * Get the min time datum for a time type.
+ *
+ * Note that the min is not the same the actual "min" of the underlying
+ * storage for date and timestamps.
+ */
+Datum
+ts_time_datum_get_min(Oid timetype)
+{
+	switch (timetype)
+	{
+		case DATEOID:
+			return DateADTGetDatum(TS_DATE_MIN);
+		case TIMESTAMPOID:
+			return TimestampGetDatum(TS_TIMESTAMP_MIN);
+		case TIMESTAMPTZOID:
+			return TimestampTzGetDatum(TS_TIMESTAMP_MIN);
+		case INT2OID:
+			return Int16GetDatum(PG_INT16_MIN);
+		case INT4OID:
+			return Int32GetDatum(PG_INT32_MIN);
+		case INT8OID:
+			return Int64GetDatum(PG_INT64_MIN);
+		default:
+			break;
+	}
+
+	return ts_time_datum_get_min(coerce_to_time_type(timetype));
+}
+
+/*
+ * Get the end time datum for a time type.
+ *
+ * Note that the end is not the same as "max" (hence not named max). The end
+ * is exclusive for date and timestamps, and might not be the same as the max
+ * value for the underlying storage type (e.g., TIMESTAMP_END is before
+ * PG_INT64_MAX). Instead, the max value for dates and timestamps represent
+ * -Infinity and +Infinity.
+ */
+Datum
+ts_time_datum_get_end(Oid timetype)
+{
+	switch (timetype)
+	{
+		case DATEOID:
+			return DateADTGetDatum(TS_DATE_END);
+		case TIMESTAMPOID:
+			return TimestampGetDatum(TS_TIMESTAMP_END);
+		case TIMESTAMPTZOID:
+			return TimestampTzGetDatum(TS_TIMESTAMP_END);
+		case INT2OID:
+			return Int16GetDatum(PG_INT16_MAX);
+		case INT4OID:
+			return Int32GetDatum(PG_INT32_MAX);
+		case INT8OID:
+			return Int64GetDatum(PG_INT64_MAX);
+		default:
+			break;
+	}
```

Comment:
I am not changing anything from what was there before. Actually, previously, there was no consistent way of dealing with this at all, which meant it was on a case-by-case basis or not handled; which led to several bugs or simply failures when close to boundaries.

The main issue is that we map several types to a common internal type (int64) and this means we have to have a representation of infinity and valid time ranges while representing in internal time. Since integers don't have a native representation of "end' or +- infinity, we have to represent it in some way. The alternative is special-casing this at every place in the code where we deal with this values, which is IMO a much bigger risk for bugs compared to having a representation in one place that can be covered by tests.

If we were to distinctly represent end and infinity with integer values, we'd have to reduce the range of supported values. I am not sure that is a good idea either (although we did previously fail when trying to insert integers close to max, due to one of the bugs).